### PR TITLE
add "save event to calendar" functionality on calendar page

### PR DESCRIPTION
**Pull request by wirednkod**
A small calendar icon is added to each event in order to allow user to download ics file and add it to calendar Screenshot 2020-11-25 at 12 12 02 AM

### CODE REVIEW

File: packages/page-calendar/src/DayItem.tsx [hunk 1]
```diff
@@ -46,62 +50,116 @@ function DayItem ({ className, item: { date, info, type } }: Props): React.React
                     : ['treasurySpend'].includes(type)
                       ? <div className='itemLink'><a href='#/treasury'>{t<string>('via Treasury')}</a></div>
                       : undefined;
+      let s = '';
 
       switch (type) {
         case 'councilElection':
-          return <><div className='itemDesc'>{t<string>('Election of new council candidates')}</div>{typeLink}</>;
-
+          s = 'Election of new council candidates';
```
Comment (jacogr) [inline packages/page-calendar/src/DayItem.tsx:57]: So on these we need to bring back the `t<string>('Election of new council candidates')` wrapper - that allows translation. Without it (or being dynamic), the i18n functionality is lost. The output can still be uses (i.e. the result from the `t(...)`) in the iCal, it is just a string.

TL;DR For all of these, just remove the fragment/div wrappers, keep the `t(...)`

File: packages/page-calendar/src/DayItem.tsx [hunk 2]
```diff
@@ -46,62 +50,116 @@ function DayItem ({ className, item: { date, info, type } }: Props): React.React
...
+          s = 'Acceptance of new members and bids';
+          break;
 
         default:
           return assertUnreachable(type);
       }
+
+      setDescription(id ? s.replaceAll('{{id}}', id) : s);
+
+      return id ? <><div className='itemDesc'>{t<string>(s, { replace: { id } })}</div>{typeLink}</> : (<><div className='itemDesc'>{s}</div>{typeLink}</>);
```
Comment (jacogr) [inline packages/page-calendar/src/DayItem.tsx:113]: As above, this is fine, but we have lost the much-needed i18n

File: packages/page-calendar/src/DayItem.tsx [hunk 3]
```diff
@@ -46,62 +50,116 @@ function DayItem ({ className, item: { date, info, type } }: Props): React.React
...
     },
     [info, t, type]
   );
 
+  function exportToCal (fileName: string, date: Date): void {
```
Comment (jacogr) [inline packages/page-calendar/src/DayItem.tsx:118]: Extract this function to file-level - as it stands it gets re-created each render.

File: packages/page-calendar/src/DayItem.tsx [hunk 4]
```diff
@@ -46,62 +50,116 @@ function DayItem ({ className, item: { date, info, type } }: Props): React.React
...
+
+    anchor.href = window.URL.createObjectURL(data);
+    anchor.download = fileNameIcs;
+    document.body.appendChild(anchor);
+    anchor.click();
+    document.body.removeChild(anchor);
+    window.URL.revokeObjectURL(anchor.href);
+  }
+
+  function calendarIcon (date: Date): React.ReactNode {
```
Comment (jacogr) [inline packages/page-calendar/src/DayItem.tsx:148]: Also extract to file-level.

File: packages/page-calendar/src/DayItem.tsx [hunk 5]
```diff
@@ -46,62 +50,116 @@ function DayItem ({ className, item: { date, info, type } }: Props): React.React
...
+    if (date) {
+      return (<Button className='exportCal'
+        icon='calendar-plus'
+        onClick={() => exportToCal(description, date)} />);
```
Comment (jacogr) [inline packages/page-calendar/src/DayItem.tsx:152]: This we would want to do as a `useCallback`

**Review by jacogr — CHANGES_REQUESTED** [review]
Thanks an absolute million for this - some small comments, but generally the functionality and look seems spot-on.

File: packages/page-calendar/src/DayItem.tsx [hunk 6]
```diff
@@ -46,62 +50,116 @@ function DayItem ({ className, item: { date, info, type } }: Props): React.React
...
+    } else {
+      return null;
+    }
+  }
+
   return (
     <div className={className}>
       <div className='itemTime'>{date.toLocaleTimeString().split(':').slice(0, 2).join(':')}</div>
       {desc}
+      {calendarIcon(date)}
```
Comment (jacogr) [inline packages/page-calendar/src/DayItem.tsx:162]: Sorry, this comment got lost (for some obscure reason). The above can just simply be (without extracting the function to file-level) -

```suggestion
      {date && (
        <Button
          className='exportCal'
          icon='calendar-plus'
          onClick={_exportCal}
      )}
```

(It assumes we have a `useCallback` with the name `_exportCal`)

**Comment by wirednkod** [comment]
> Thanks an absolute million for this - some small comments, but generally the functionality and look seems spot-on.

Its a pleasure :) Thanks for the feedback

**Review by jacogr — APPROVED** [review]
Absolutely brilliant! I'll be making extensive use of this, as are others.

**Comment by Swader** [comment]
@wirednkod this is great, can you let us know a DOT address? I'd like to tip you for this work.

**Comment by wirednkod** [comment]
@Swader Thank you for the offer but there is really no need for a tip, mate. It was done for pleasure - rather than payment/tip. Was interested on having a look on the code for an interview with Parity and stumbled upon this one - seemed really easy and fast to create - so why not  ¯\_(ツ)_/¯

**Comment by wirednkod** [comment]
@jacogr I have found various 3rd party npms that are making the export easier to download for Google/Apple/Outlook etc, etc. I could expand this feature to that - if you are ok with it :) Do you prefer custom code or 3rd parties are fine as well?

**Comment by Swader** [comment]
> @Swader Thank you for the offer but there is really no need for a tip, mate. It was done for pleasure - rather than payment/tip. Was interested on having a look on the code for an interview with Parity and stumbled upon this one - seemed really easy and fast to create - so why not ¯_(ツ)_/¯

Admirable, but it's really commonplace: https://ipfs.io/ipns/polkadot.dotapps.io/#/treasury/tips - we really do like to support our supporters, especially since it's funded by the chain itself. If you change your mind, the offer stands, all I need is an address :)

**Comment by jacogr** [comment]
Really depends on the value the npms actually add - if it is left-pad like, then I would rather bring it in. 

Each dep ends up with having to be managed for the next 20 years and each bring in their level of cruft and so it just adds to the pile. Some weeks can't keep up with security advisories, so generally unless it is really novel, try to steer clear from too many external deps, there are too many already bringing in their own pile... 

Trying to clean up here as well, but current efforts in that regard actually sits in other repos, so will get here for "spring cleaning" :)

**Comment by wirednkod** [comment]
> Each dep ends up with having to be managed for the next 20 years and each bring in their level of cruft and so it just adds to the pile. Some weeks can't keep up with security advisories, so generally unless it is really novel work, try to steer clear from too many external deps, there are too many already bringing in their own pile...

Makes sense :) I'll have a look to wrap it in custom code in order to avoid dependencies. Cheers

**Comment by wirednkod** [comment]
> > @Swader Thank you for the offer but there is really no need for a tip, mate. It was done for pleasure - rather than payment/tip. Was interested on having a look on the code for an interview with Parity and stumbled upon this one - seemed really easy and fast to create - so why not ¯_(ツ)_/¯
> 
> Admirable, but it's really commonplace: https://ipfs.io/ipns/polkadot.dotapps.io/#/treasury/tips - we really do like to support our supporters, especially since it's funded by the chain itself. If you change your mind, the offer stands, all I need is an address :)

Well.. if its a commonplace then here is an address:
11e2DNjBooKKoiXnt9Mw9ePi3PSDF9bUvh1ntYctAfUUoYD

Cheers
